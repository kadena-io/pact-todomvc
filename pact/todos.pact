;;
;; todos smart contract
;;

;; admin keyset definition
(define-keyset 'todo-admin-keyset
  (read-keyset "todo-admin-keyset"))

;; todos module, administered by keyset
(module todos 'todo-admin-keyset
  " Smart contract module for TODO-MVC pact app.   \
  \ Tables:                                        \
  \ todo -- holds todo entries"

  ;; todo schema and table
  (defschema todo
    "Row type for todos."
     title:string
     completed:bool
     deleted:bool
     )

  (deftable todo-table:{todo})

  ;;
  ;; API functions
  ;;

  (defun new-todo (id title)
    "Create new todo with ENTRY and DATE."
    (insert todo-table id {
      "title": title,
      "completed": false,
      "deleted": false })
  )

  (defun toggle-todo-status (id:string)
    "Toggle completed status flag for todo at ID."
    (with-read todo-table id {
      "completed":= state
      }
      (update todo-table id {
        "completed": (not state) })))

  (defun edit-todo (id:string title)
    "Update todo ENTRY at ID."
    (update todo-table id {
      "title": title }))

  (defun delete-todo (id:string)
    "Delete todo title at ID (by setting deleted flag)."
    (update todo-table id {
      "deleted": true }))

  (defun read-todo:object (id:string)
    "Read a single todo"
    (+ {'id: id} (read todo-table id)))

  (defun read-todos:[object{todo}] ()
    "Read all todos."
    (map (read-todo) (keys todo-table)))
)

(create-table todo-table)
;done
