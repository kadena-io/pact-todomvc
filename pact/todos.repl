(env-keys ["todoadmin"])
(env-data { "todo-admin-keyset":
  { "keys": ["todoadmin"], "pred": "keys-all" } })

(begin-tx)
(load "todos.pact")
(commit-tx)

(typecheck "todos")

(use todos)

;;Create a new todo
(new-todo "1" "wash")

;;Test if todo was created
(expect "Returns the todo object with id-1" {"completed": false, "deleted": false, "id": "1", "title": "wash"} (read-todo "1"))

;;Fail if todo to be edited does not exist.
(expect-failure "No todo exists with id of 2." (edit-todo "2" "clean"))
