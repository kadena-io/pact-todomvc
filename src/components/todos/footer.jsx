import * as React from "react";

export class TodoFooter extends React.PureComponent {
  render() {
    var pluralize = function(count, word) {
      return count === 1 ? word : word + "s";
    };
    var activeTodoWord = pluralize(this.props.count, "item");
    var clearButton = null;

    if (this.props.completedCount > 0) {
      clearButton = (
        <button
          className="clear-completed"
          onClick={this.props.onClearCompleted}
        >
          Clear completed
        </button>
      );
    }

    var nowShowing = this.props.nowShowing;
    return (
      <footer className="footer">
        <span className="todo-count">
          <strong>{this.props.count}</strong> {activeTodoWord} left
        </span>
        <ul className="filters">
          <li>
            <a href="#/" className={{ selected: nowShowing === "all" }}>
              All
            </a>
          </li>{" "}
          <li>
            <a
              href="#/active"
              className={{ selected: nowShowing === "active" }}
            >
              Active
            </a>
          </li>{" "}
          <li>
            <a
              href="#/completed"
              className={{ selected: nowShowing === "completed" }}
            >
              Completed
            </a>
          </li>
        </ul>
        {clearButton}
      </footer>
    );
  }
}
