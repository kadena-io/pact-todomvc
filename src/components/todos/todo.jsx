import * as React from "react";

var ESCAPE_KEY = 27;
var ENTER_KEY = 13;

export class TodoItem extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = { editText: this.props.todo.title };
  }

  handleSubmit(event) {
    var val = this.state.editText.trim();
    if (val) {
      this.props.onSave(val);
      this.setState({ editText: val });
    } else {
      this.props.onDestroy();
    }
  }

  handleEdit() {
    this.props.onEdit();
    this.setState({ editText: this.props.todo.title });
  }

  handleKeyDown(event) {
    if (event.which === ESCAPE_KEY) {
      this.setState({ editText: this.props.todo.title });
      this.props.onCancel(event);
    } else if (event.which === ENTER_KEY) {
      this.handleSubmit(event);
    }
  }

  handleChange(event) {
    if (this.props.editing) {
      this.setState({ editText: event.target.value });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      nextProps.todo !== this.props.todo ||
      nextProps.editing !== this.props.editing ||
      nextState.editText !== this.state.editText
    );
  }
  componentDidUpdate(prevProps) {
    if (!prevProps.editing && this.props.editing) {
      var node = React.findDOMNode(this.refs.editField);
      node.focus();
      node.setSelectionRange(node.value.length, node.value.length);
    }
  }

  render() {
    return (
      <li
        className={{
          completed: this.props.todo.completed,
          editing: this.props.editing
        }}
      >
        <div className="view">
          <input
            className="toggle"
            type="checkbox"
            checked={this.props.todo.completed}
            onChange={this.props.onToggle}
          />
          <label onDoubleClick={this.handleEdit}>{this.props.todo.title}</label>
          <button className="destroy" onClick={this.props.onDestroy} />
        </div>
        <input
          ref="editField"
          className="edit"
          value={this.props.editText}
          onBlur={this.handleSubmit}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
        />
      </li>
    );
  }
}
