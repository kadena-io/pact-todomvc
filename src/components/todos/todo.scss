$complete: #900;
$correct: #09c;

.todo {
  display: flex;
  margin-bottom: 0.5em;

  button {
    appearance: none;
    padding: 0.5em;
    border-radius: 3px;
    color: #333;
    background: #fff;
    border: none;
    cursor: pointer;
  }

  .check button:hover {
    background: #09c;
    color: #fff;
  }

  .remove button:hover {
    background: #c00;
    color: #fff;
  }

  .entry {
    flex: 1;
    text-align: left;
    margin: 0 0.2em;

    input {
      color: #000;
      border: 1px solid transparent;
      transition: all 0.1s ease-in-out;
      border-radius: 3px;
      width: 100%;
      height: 100%;
      padding: 0 0.5em;
      font-size: 0.9em;

      &:hover,
      &.empty {
        border-color: #ccc;
      }

      &:focus {
        border-color: #999;
      }
    }
  }

  &.new {
    .entry {
      height: 1.5em;
    }
  }
  &.complete {
    color: $complete;
    .entry {
      text-decoration: strikethrough;
    }
  }
}
