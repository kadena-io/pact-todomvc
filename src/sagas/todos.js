import { call, put } from 'redux-saga/effects';

import sendPactCommand from '../utils/send-pact-command';

export const FETCH_TODOS = 'pact-todomvc/todos/FETCH_TODOS';
export const FETCH_TODOS_REQUEST = 'pact-todomvc/todos/FETCH_TODOS_REQUEST';
export const FETCH_TODOS_SUCCEEDED = 'pact-todomvc/todos/FETCH_TODOS_SUCCEEDED';
export const FETCH_TODOS_FAILED = 'pact-todomvc/todos/FETCH_TODOS_FAILED';

export const SAVE_NEW_TODO = 'pact-todomvc/todos/SAVE_NEW_TODO';
export const SAVE_NEW_TODO_REQUEST = 'pact-todomvc/todos/SAVE_NEW_TODO_REQUEST';
export const SAVE_NEW_TODO_SUCCEEDED = 'pact-todomvc/todos/SAVE_NEW_TODO_SUCCEEDED';
export const SAVE_NEW_TODO_FAILED = 'pact-todomvc/todos/SAVE_NEW_TODO_FAILED';

export const REMOVE_TODO = 'pact-todomvc/todos/REMOVE_TODO';
export const REMOVE_TODO_REQUEST = 'pact-todomvc/todos/REMOVE_TODO_REQUEST';
export const REMOVE_TODO_SUCCEEDED = 'pact-todomvc/todos/REMOVE_TODO_SUCCEEDED';
export const REMOVE_TODO_FAILED = 'pact-todomvc/todos/REMOVE_TODO_FAILED';

export const UPDATE_TODO = 'pact-todomvc/todos/UPDATE_TODO';
export const UPDATE_TODO_REQUEST = 'pact-todomvc/todos/UPDATE_TODO_REQUEST';
export const UPDATE_TODO_SUCCEEDED = 'pact-todomvc/todos/UPDATE_TODO_SUCCEEDED';
export const UPDATE_TODO_FAILED = 'pact-todomvc/todos/UPDATE_TODO_FAILED';

export const TOGGLE_TODO_STATE = 'pact-todomvc/todos/TOGGLE_TODO_STATE';
export const TOGGLE_TODO_STATE_REQUEST = 'pact-todomvc/todos/TOGGLE_TODO_STATE_REQUEST';
export const TOGGLE_TODO_STATE_SUCCEEDED = 'pact-todomvc/todos/TOGGLE_TODO_STATE_SUCCEEDED';
export const TOGGLE_TODO_STATE_FAILED = 'pact-todomvc/todos/TOGGLE_TODO_STATE_FAILED';

export const CHANGE_EDIT_STATUS = 'pact-todomvc/todos/CHANGE_EDIT_STATUS';
export const CHANGE_TODO_ENTRY = 'pact-todomvc/todos/CHANGE_TODO_ENTRY';
export const UPDATE_NEW_TODO_FIELD = 'pact-todomvc/todos/UPDATE_NEW_TODO_FIELD';

const initialState = {
  todosIsLoading: false,
  todosError: null,
  todos: [],
  newTodo: '',
  editStatus: false,
  editedTodo: null,
};

let workingTodo;
export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case FETCH_TODOS_REQUEST:
      return { ...state, todosIsLoading: true };

    case FETCH_TODOS_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: action.todos
          .sort((a, b) => a.id - b.id)
          .map(todo => ({ ...todo, editStatus: false })),
        todosIsLoading: false,
      };

    case FETCH_TODOS_FAILED:
      return {
        ...state,
        todosError: action.error,
        todos: [],
        todosIsLoading: false,
      };

    case SAVE_NEW_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case SAVE_NEW_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: [...state.todos, action.newTodo].sort((a, b) => a.id - b.id),
        todosIsLoading: false,
        newTodo: '',
      };

    case SAVE_NEW_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case REMOVE_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case REMOVE_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: state.todos.filter(todo => todo.id !== action.id),
        todosIsLoading: false,
        newTodo: '',
      };

    case REMOVE_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case UPDATE_TODO_REQUEST:
      return { ...state, todosIsLoading: true };

    case UPDATE_TODO_SUCCEEDED:
      return {
        ...state,
        todosError: null,
        todos: [
          ...state.todos.filter(todo => todo.id !== action.todo.id),
          { ...action.todo, editStatus: false },
        ].sort((a, b) => a.id - b.id),
        todosIsLoading: false,
      };

    case UPDATE_TODO_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case TOGGLE_TODO_STATE_REQUEST:
      return { ...state, todosIsLoading: true };

    case TOGGLE_TODO_STATE_SUCCEEDED:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.state = action.state;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
        todosError: null,
        todosIsLoading: false,
      };

    case TOGGLE_TODO_STATE_FAILED:
      return {
        ...state,
        todosError: action.error,
        todosIsLoading: false,
      };

    case CHANGE_TODO_ENTRY:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.entry = action.entry;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
      };

    case CHANGE_EDIT_STATUS:
      workingTodo = state.todos.find(todo => todo.id === action.id);
      workingTodo.editStatus = !workingTodo.editStatus;
      return {
        ...state,
        todos: [...state.todos.filter(todo => todo.id !== action.id), workingTodo].sort(
          (a, b) => a.id - b.id
        ),
      };

    case UPDATE_NEW_TODO_FIELD:
      return { ...state, newTodo: action.newTodo };

    default:
      return state;
  }
}

export function updateNewTodoField(newTodo) {
  return { type: UPDATE_NEW_TODO_FIELD, newTodo };
}

export function fetchTodos() {
  return { type: FETCH_TODOS };
}

export function saveNewTodo(entry, date) {
  return { type: SAVE_NEW_TODO, entry, date };
}

export function changeEntry(id, entry, date) {
  return {
    type: CHANGE_TODO_ENTRY,
    id,
    entry,
    date,
  };
}

export function updateTodo(todo) {
  return { type: UPDATE_TODO, todo };
}

export function toggleState(id, state) {
  return { type: TOGGLE_TODO_STATE, id, state };
}

export function removeTodo(id) {
  return { type: REMOVE_TODO, id };
}

export function changeEditStatus(id) {
  return { type: CHANGE_EDIT_STATUS, id };
}

export function* fetchTodosSaga() {
  yield put({ type: FETCH_TODOS_REQUEST });
  try {
    const todos = yield call(sendPactCommand, '(todos.read-todos)');
    yield put({ type: FETCH_TODOS_SUCCEEDED, todos });
  } catch (error) {
    yield put({ type: FETCH_TODOS_FAILED, error });
  }
}

export function* toggleTodoStateSaga({ id, state }) {
  yield put({ type: TOGGLE_TODO_STATE_REQUEST });
  try {
    yield call(sendPactCommand, `(todos.toggle-todo-status ${id})`);
    yield put({ type: TOGGLE_TODO_STATE_SUCCEEDED, id, state });
  } catch (error) {
    yield put({ type: TOGGLE_TODO_STATE_FAILED, error });
  }
}

export function* saveNewTodoSaga({ entry, date }) {
  yield put({ type: SAVE_NEW_TODO_REQUEST });
  try {
    const newTodo = yield call(
      sendPactCommand,
      `(todos.new-todo ${JSON.stringify(entry)} ${JSON.stringify(date)})`
    );
    yield put({ type: SAVE_NEW_TODO_SUCCEEDED, newTodo });
  } catch (error) {
    yield put({ type: SAVE_NEW_TODO_FAILED, error });
  }
}

export function* removeTodoSaga({ id }) {
  yield put({ type: REMOVE_TODO_REQUEST, id });
  try {
    yield call(sendPactCommand, `(todos.delete-todo ${id})`);
    yield put({ type: REMOVE_TODO_SUCCEEDED, id });
  } catch (error) {
    yield put({ type: REMOVE_TODO_FAILED, error });
  }
}

export function* updateTodoSaga({ todo }) {
  yield put({ type: UPDATE_TODO_REQUEST, todo });
  try {
    yield call(
      sendPactCommand,
      `(todos.edit-todo ${todo.id} ${JSON.stringify(todo.entry)}  ${JSON.stringify(todo.date)})`
    );
    yield put({ type: UPDATE_TODO_SUCCEEDED, todo });
  } catch (error) {
    yield put({ type: UPDATE_TODO_FAILED, error });
  }
}
