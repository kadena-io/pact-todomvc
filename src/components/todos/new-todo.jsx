import * as React from 'react';
import './todo.scss';

export const NewTodo = ({ saveNewTodo }) => {
  let entry='';
  let date;

  const handleDate = e => {
    e.preventDefault();
    date = e.currentTarget.value;
  };

  const handleEntry = e => {
    e.preventDefault();
    entry = e.currentTarget.value;
  };

  const entryKeyDown = e => {
    if (e.keyCode === 13) {
      let newDate = new Date();
      date = date || newDate.toISOString().slice(0,10);
      console.log(entry, date)
      saveNewTodo(entry, date);
      entry='';
    }
  };

  return (
    <div className="todo new">
      <div className="entry">
        <label>New Task</label>
        <input id= "new-todo" type="text" className="empty" onChange={handleEntry} onKeyDown={entryKeyDown} placeholder="New Todoâ€¦" />
      </div>
      <div className="date">
        <label>Due Date</label>
        <input id="due-date" type="date" onChange={handleDate} onKeyDown={entryKeyDown} placeholder="Enter Due Date..." />
      </div>
    </div>
  );
};
