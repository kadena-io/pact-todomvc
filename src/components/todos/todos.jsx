import * as React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Loader from 'react-loader-spinner';

import { fetchTodos } from '../../sagas/todos';

import { Todo } from './todo';

class TodosComponent extends React.PureComponent {
  constructor() {
    super();
    this.fetchTodos = this.fetchTodos.bind(this);
  }

  componentDidMount() {
    this.fetchTodos();
  }

  fetchTodos() {
    this.props.fetchTodos();
  }

  render() {
    const { todos, todosIsLoading, todosError } = this.props;

    if (todosIsLoading) {
      return (
        <div className="todos">
          <h1>Todos</h1>
          <Loader type="TailSpin" color="#3cf" />
        </div>
      );
    } else if (todosError !== null) {
      return (
        <div className="todos">
          <h1>Todos</h1>
          <div className="error">Error loading todos</div>
          <p>
            <Link className="btn" to="/" onClick={this.fetchTodos}>
              Reload
            </Link>
          </p>
        </div>
      );
    } else {
      return (
        <div className="todos">
          <h1>Todos</h1>'
          <div className="show-completed">
            <input type="checkbox" /> Show Completed Items
          </div>
          <div className="todos-list">
            {todos.map((todo, i) => (
              <Todo {...todo} key={i} />
            ))}
          </div>
          <p>
            <Link className="btn" to="/">
              New Todo
            </Link>
          </p>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  /*
    todosIsLoading: false,
    todosError: null,
    todos: []
  */
  return { ...state.todos };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchTodos: () => {
      dispatch(fetchTodos());
    },
    saveNewTodo: newTodo => {
      dispatch(saveNewTodo(newTodo));
    },
    removeTodo: id => {
      dispatch(removeTodo(id));
    },
  };
};

export const Todos = connect(
  mapStateToProps,
  mapDispatchToProps
)(TodosComponent);
