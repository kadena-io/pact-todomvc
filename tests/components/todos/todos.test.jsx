import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { TodosComponent } from '../../../src/components/todos/todos';
import { NewTodo } from '../../../src/components/todos/new-todo';
import { Todo } from '../../../src/components/todos/todo';

const todo = {
  id: 1,
  entry: 'A Todo',
  state: 'active',
};

const todosProps = {
  todosIsLoading: false,
  todosError: null,
  todos: [todo],
  newTodo: '',
  editedTodo: null,
  fetchTodos: jest.fn(),
  saveNewTodo: jest.fn(),
  removeTodo: jest.fn(),
  updateTodo: jest.fn(),
  toggleState: jest.fn(),
  changeEntry: jest.fn(),
  updateNewTodoField: jest.fn(),
};

describe('Todos Component', () => {
  describe('Pass-thru Functions', () => {
    const component = shallow(<TodosComponent {...todosProps} />).instance();

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('should render a todos list', () => {
      const snapshot = renderer.create(<TodosComponent {...todosProps} />);
      const tree = snapshot.toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('should call fetchTodos() on componentDidMount()', () => {
      jest.spyOn(component, 'fetchTodos');
      component.componentDidMount();
      expect(component.fetchTodos).toHaveBeenCalled();
    });

    test('should call props.fetchTodos() on this.fetchTodos()', () => {
      component.fetchTodos();
      expect(todosProps.fetchTodos).toHaveBeenCalled();
    });

    test('should call props.saveNewTodo() on this.saveNewTodo()', () => {
      component.saveNewTodo(todo.entry);
      expect(todosProps.saveNewTodo).toHaveBeenCalledWith(todo.entry);
    });

    test('should call props.removeTodo() on this.onRemoveTodo()', () => {
      component.onRemoveTodo(todo.id);
      expect(todosProps.removeTodo).toHaveBeenCalledWith(todo.id);
    });

    test('should call props.updateTodo() on this.onUpdateTodo()', () => {
      component.onUpdateTodo(todo);
      expect(todosProps.updateTodo).toHaveBeenCalledWith(todo);
    });

    test('should call props.toggleState() on this.onToggleState()', () => {
      component.onToggleState(todo.id, todo.state);
      expect(todosProps.toggleState).toHaveBeenCalledWith(todo.id, todo.state);
    });

    test('should call props.changeEntry() on this.onChangeEntry()', () => {
      component.onChangeEntry(todo.id, todo.entry);
      expect(todosProps.changeEntry).toHaveBeenCalledWith(todo.id, todo.entry);
    });
  });

  describe('Render', () => {
    describe('Top Level', () => {
      test('should render an <h1>Todos</div> and a root element with classname "todos"', () => {
        const component = shallow(<TodosComponent {...todosProps} />);
        expect(component.find('h1').text()).toEqual('Todos');
        expect(component.find('div .todos-list').exists()).toBeTruthy();
        expect(component.find(NewTodo).exists()).toBeTruthy();
        expect(component.find(Todo).exists()).toBeTruthy();
      });

      test('should render a <div className="error" /> if todoError !== null', () => {
        const component = shallow(
          <TodosComponent {...todosProps} todosError="Error loading todos" />
        );
        expect(component.find('.error').exists()).toBeTruthy();
      });
    });

    describe('Child Handlers', () => {
      const component = shallow(<TodosComponent {...todosProps} />);
      const instance = component.instance();
      const todo = component.find(Todo);

      test('NewTodo.saveNewTodo should call saveNewTodo()', () => {
        const newTodo = component.find(NewTodo);
        expect(newTodo.prop('saveNewTodo')).toEqual(instance.saveNewTodo);
      });

      test('Todo.onRemove should call onRemoveTodo()', () => {
        expect(todo.prop('onRemove')).toEqual(instance.onRemoveTodo);
      });

      test('Todo.onChangeEntry should call onChangeEntry()', () => {
        expect(todo.prop('onChangeEntry')).toEqual(instance.onChangeEntry);
      });

      test('Todo.onUpdate should call onUpdateTodo()', () => {
        expect(todo.prop('onUpdate')).toEqual(instance.onUpdateTodo);
      });

      test('Todo.onToggleState should call onToggleState()', () => {
        expect(todo.prop('onToggleState')).toEqual(instance.onToggleState);
      });
    });
  });
});
