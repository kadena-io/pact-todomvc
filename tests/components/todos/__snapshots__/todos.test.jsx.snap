// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Component Pass-thru Functions should render a todos list 1`] = `
<div
  className="todos"
>
  <h1>
    Todos
  </h1>
  <div
    className="todos-list"
  >
    <div
      className="todo new"
    >
      <div
        className="entry"
      >
        <input
          className="empty"
          onKeyDown={[Function]}
          placeholder="New Todo…"
          type="text"
        />
      </div>
    </div>
    <div
      className="todo active"
    >
      <div
        className="check"
      >
        <button
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-circle fa-w-16 "
            data-icon="circle"
            data-prefix="far"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
      <div
        className="entry"
      >
        <input
          className=""
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Enter Todo…"
          type="text"
          value="A Todo"
        />
      </div>
      <div
        className="remove"
      >
        <button
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-times fa-w-11 "
            data-icon="times"
            data-prefix="fas"
            role="img"
            style={Object {}}
            viewBox="0 0 352 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
